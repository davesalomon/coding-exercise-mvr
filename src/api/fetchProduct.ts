import moize from "moize";
import config from "../config";

type VenueData = {
  id: number;
  name: string;
  description: string;
  city: string;
  country_code: string;
};

export type ArtistData = {
  id: number;
  name: string;
  description: string;
  country_code: string;
  website_url: string;
  facebook_url: string;
  instagram_url: string;
  twitter_url: string;
  image_url: string;
  artist_background_colour_code: string;
  artist_text_colour_code: string;
  artist_accent_colour_code: string;
};

type GenreData = {
  id: number;
  name: string;
  description: string | null;
};

export type DetailedProduct = {
  id: number;
  title: string;
  description: string;
  image_hero_url: string;
  image_landscape_url: string;
  accent_colour_code: string;
  background_colour_code: string;
  text_colour_code: string;
  production_type_code: "performance" | "original";
  price: string;
  venue: VenueData;
  artist: ArtistData;
  genre: GenreData;
};

const fetchProduct = async (id: string): Promise<DetailedProduct> => {
  const product = await fetch(
    config.api.productDetails.replace("{id}", id.toString())
  );

  if (product.ok) {
    return (await product.json()).result as DetailedProduct;
  }

  throw new Error(
    `Failed to fetch products - ${product.status} ${product.statusText}`
  );
};

export default moize(fetchProduct);
