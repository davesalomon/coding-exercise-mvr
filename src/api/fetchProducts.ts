import config from "../config";
import fetchProduct from "./fetchProduct";

export type ProductDetails = {
  accent_colour_code: string;
  background_colour_code: string;
  description: string;
  id: number;
  image_hero_url: string;
  image_landscape_url: string;
  price?: string;
  production_type_code: "performance" | "original";
  text_colour_code: string;
  title: string;
};

export type Product = {
  feature_order: number;
  is_free: boolean;
  product: ProductDetails;
};

const fetchProducts = async (): Promise<Product[]> => {
  const products = await fetch(config.api.productsList);
  if (products.ok) {
    const result = (await products.json()).result.releases as Product[];
    result.sort((a, b) => a.feature_order - b.feature_order);

    const additionalProductDetailsPromises: any[] = (
      await Promise.all(
        result.map((product) => {
          // I wouldn't do this in real life -- it's a backend fix that needs to happen.
          return new Promise(async (resolve) => {
            if (
              !product.is_free &&
              typeof product.product.price === "undefined"
            ) {
              const detailedProductInfo = await fetchProduct(
                product.product.id.toString()
              );
              resolve({
                id: product.product.id,
                price: detailedProductInfo.price,
              });
            }
            return resolve();
          });
        })
      )
    ).filter((a) => a);

    return result.map((product) => {
      const additionalProduct = additionalProductDetailsPromises.find(
        (p) => p.id === product.product.id
      );

      return {
        ...product,
        product: {
          ...product.product,
          ...additionalProduct,
        },
      };
    });
  }

  throw new Error(
    `Failed to fetch products - ${products.status} ${products.statusText}`
  );
};

export default fetchProducts;
