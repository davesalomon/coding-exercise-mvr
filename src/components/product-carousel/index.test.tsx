import React from "react";
import { render } from "@testing-library/react";
import ProductsCarousel from "./";
import { ProductDetails } from "../../api/fetchProducts";

jest.mock("react-router-dom", () => ({
  Link: jest.fn(() => "<Link />"),
}));

const generateProduct = (i: number): ProductDetails => ({
  accent_colour_code: "#f00",
  background_colour_code: "#f0f",
  description: "",
  id: i,
  image_hero_url: "https://example.com/foo.jpg",
  image_landscape_url: "",
  price: i.toFixed(2),
  production_type_code: "performance",
  text_colour_code: "#00f",
  title: `Product ${i}`,
});

test("renders a basic default layout without a title", () => {
  const result = render(
    <ProductsCarousel
      products={[
        { feature_order: 1, is_free: false, product: generateProduct(1) },
        { feature_order: 1, is_free: false, product: generateProduct(2) },
        { feature_order: 1, is_free: false, product: generateProduct(3) },
        { feature_order: 1, is_free: false, product: generateProduct(4) },
        { feature_order: 1, is_free: false, product: generateProduct(5) },
        { feature_order: 1, is_free: false, product: generateProduct(6) },
        { feature_order: 1, is_free: false, product: generateProduct(7) },
        { feature_order: 1, is_free: false, product: generateProduct(8) },
        { feature_order: 1, is_free: false, product: generateProduct(9) },
        { feature_order: 1, is_free: false, product: generateProduct(10) },
        { feature_order: 1, is_free: false, product: generateProduct(11) },
        { feature_order: 1, is_free: false, product: generateProduct(12) },
      ]}
    />
  );

  expect(result).toMatchSnapshot();
});
