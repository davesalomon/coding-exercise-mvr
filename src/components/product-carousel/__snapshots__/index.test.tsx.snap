// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a basic default layout without a title 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="makeStyles-container-1"
      >
        <div
          class="makeStyles-root-3"
        >
          <div
            class="makeStyles-container-4"
          >
            <div
              class="makeStyles-productTile-13"
            >
              <div
                class="makeStyles-productImage-12"
              />
              <div>
                <h6
                  class="MuiTypography-root makeStyles-productName-10 MuiTypography-h6"
                >
                  Product 1
                </h6>
                <div
                  class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall"
                  >
                    £1.00
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root makeStyles-productDescription-11 MuiTypography-body1"
              />
              &lt;Link /&gt;
            </div>
            <div
              class="makeStyles-productTile-20"
            >
              <div
                class="makeStyles-productImage-19"
              />
              <div>
                <h6
                  class="MuiTypography-root makeStyles-productName-17 MuiTypography-h6"
                >
                  Product 2
                </h6>
                <div
                  class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall"
                  >
                    £2.00
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root makeStyles-productDescription-18 MuiTypography-body1"
              />
              &lt;Link /&gt;
            </div>
            <div
              class="makeStyles-productTile-25"
            >
              <div
                class="makeStyles-productImage-24"
              />
              <div>
                <h6
                  class="MuiTypography-root makeStyles-productName-22 MuiTypography-h6"
                >
                  Product 3
                </h6>
                <div
                  class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall"
                  >
                    £3.00
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root makeStyles-productDescription-23 MuiTypography-body1"
              />
              &lt;Link /&gt;
            </div>
            <div
              class="makeStyles-productTile-30"
            >
              <div
                class="makeStyles-productImage-29"
              />
              <div>
                <h6
                  class="MuiTypography-root makeStyles-productName-27 MuiTypography-h6"
                >
                  Product 4
                </h6>
                <div
                  class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall"
                  >
                    £4.00
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root makeStyles-productDescription-28 MuiTypography-body1"
              />
              &lt;Link /&gt;
            </div>
            <div
              class="makeStyles-productTile-35"
            >
              <div
                class="makeStyles-productImage-34"
              />
              <div>
                <h6
                  class="MuiTypography-root makeStyles-productName-32 MuiTypography-h6"
                >
                  Product 5
                </h6>
                <div
                  class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall"
                  >
                    £5.00
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root makeStyles-productDescription-33 MuiTypography-body1"
              />
              &lt;Link /&gt;
            </div>
            <div
              class="makeStyles-productTile-40"
            >
              <div
                class="makeStyles-productImage-39"
              />
              <div>
                <h6
                  class="MuiTypography-root makeStyles-productName-37 MuiTypography-h6"
                >
                  Product 6
                </h6>
                <div
                  class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall"
                  >
                    £6.00
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root makeStyles-productDescription-38 MuiTypography-body1"
              />
              &lt;Link /&gt;
            </div>
            <div
              class="makeStyles-productTile-45"
            >
              <div
                class="makeStyles-productImage-44"
              />
              <div>
                <h6
                  class="MuiTypography-root makeStyles-productName-42 MuiTypography-h6"
                >
                  Product 7
                </h6>
                <div
                  class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall"
                  >
                    £7.00
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root makeStyles-productDescription-43 MuiTypography-body1"
              />
              &lt;Link /&gt;
            </div>
            <div
              class="makeStyles-productTile-50"
            >
              <div
                class="makeStyles-productImage-49"
              />
              <div>
                <h6
                  class="MuiTypography-root makeStyles-productName-47 MuiTypography-h6"
                >
                  Product 8
                </h6>
                <div
                  class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall"
                  >
                    £8.00
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root makeStyles-productDescription-48 MuiTypography-body1"
              />
              &lt;Link /&gt;
            </div>
            <div
              class="makeStyles-productTile-55"
            >
              <div
                class="makeStyles-productImage-54"
              />
              <div>
                <h6
                  class="MuiTypography-root makeStyles-productName-52 MuiTypography-h6"
                >
                  Product 9
                </h6>
                <div
                  class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall"
                  >
                    £9.00
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root makeStyles-productDescription-53 MuiTypography-body1"
              />
              &lt;Link /&gt;
            </div>
            <div
              class="makeStyles-productTile-60"
            >
              <div
                class="makeStyles-productImage-59"
              />
              <div>
                <h6
                  class="MuiTypography-root makeStyles-productName-57 MuiTypography-h6"
                >
                  Product 10
                </h6>
                <div
                  class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall"
                  >
                    £10.00
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root makeStyles-productDescription-58 MuiTypography-body1"
              />
              &lt;Link /&gt;
            </div>
            <div
              class="makeStyles-productTile-65"
            >
              <div
                class="makeStyles-productImage-64"
              />
              <div>
                <h6
                  class="MuiTypography-root makeStyles-productName-62 MuiTypography-h6"
                >
                  Product 11
                </h6>
                <div
                  class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall"
                  >
                    £11.00
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root makeStyles-productDescription-63 MuiTypography-body1"
              />
              &lt;Link /&gt;
            </div>
            <div
              class="makeStyles-productTile-70"
            >
              <div
                class="makeStyles-productImage-69"
              />
              <div>
                <h6
                  class="MuiTypography-root makeStyles-productName-67 MuiTypography-h6"
                >
                  Product 12
                </h6>
                <div
                  class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall"
                  >
                    £12.00
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root makeStyles-productDescription-68 MuiTypography-body1"
              />
              &lt;Link /&gt;
            </div>
          </div>
          <div
            class="makeStyles-leftArrow-7 makeStyles-arrow-6 makeStyles-disabled-5"
          />
          <div
            class="makeStyles-rightArrow-8 makeStyles-arrow-6"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="makeStyles-container-1"
    >
      <div
        class="makeStyles-root-3"
      >
        <div
          class="makeStyles-container-4"
        >
          <div
            class="makeStyles-productTile-13"
          >
            <div
              class="makeStyles-productImage-12"
            />
            <div>
              <h6
                class="MuiTypography-root makeStyles-productName-10 MuiTypography-h6"
              >
                Product 1
              </h6>
              <div
                class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall"
                >
                  £1.00
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root makeStyles-productDescription-11 MuiTypography-body1"
            />
            &lt;Link /&gt;
          </div>
          <div
            class="makeStyles-productTile-20"
          >
            <div
              class="makeStyles-productImage-19"
            />
            <div>
              <h6
                class="MuiTypography-root makeStyles-productName-17 MuiTypography-h6"
              >
                Product 2
              </h6>
              <div
                class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall"
                >
                  £2.00
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root makeStyles-productDescription-18 MuiTypography-body1"
            />
            &lt;Link /&gt;
          </div>
          <div
            class="makeStyles-productTile-25"
          >
            <div
              class="makeStyles-productImage-24"
            />
            <div>
              <h6
                class="MuiTypography-root makeStyles-productName-22 MuiTypography-h6"
              >
                Product 3
              </h6>
              <div
                class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall"
                >
                  £3.00
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root makeStyles-productDescription-23 MuiTypography-body1"
            />
            &lt;Link /&gt;
          </div>
          <div
            class="makeStyles-productTile-30"
          >
            <div
              class="makeStyles-productImage-29"
            />
            <div>
              <h6
                class="MuiTypography-root makeStyles-productName-27 MuiTypography-h6"
              >
                Product 4
              </h6>
              <div
                class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall"
                >
                  £4.00
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root makeStyles-productDescription-28 MuiTypography-body1"
            />
            &lt;Link /&gt;
          </div>
          <div
            class="makeStyles-productTile-35"
          >
            <div
              class="makeStyles-productImage-34"
            />
            <div>
              <h6
                class="MuiTypography-root makeStyles-productName-32 MuiTypography-h6"
              >
                Product 5
              </h6>
              <div
                class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall"
                >
                  £5.00
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root makeStyles-productDescription-33 MuiTypography-body1"
            />
            &lt;Link /&gt;
          </div>
          <div
            class="makeStyles-productTile-40"
          >
            <div
              class="makeStyles-productImage-39"
            />
            <div>
              <h6
                class="MuiTypography-root makeStyles-productName-37 MuiTypography-h6"
              >
                Product 6
              </h6>
              <div
                class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall"
                >
                  £6.00
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root makeStyles-productDescription-38 MuiTypography-body1"
            />
            &lt;Link /&gt;
          </div>
          <div
            class="makeStyles-productTile-45"
          >
            <div
              class="makeStyles-productImage-44"
            />
            <div>
              <h6
                class="MuiTypography-root makeStyles-productName-42 MuiTypography-h6"
              >
                Product 7
              </h6>
              <div
                class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall"
                >
                  £7.00
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root makeStyles-productDescription-43 MuiTypography-body1"
            />
            &lt;Link /&gt;
          </div>
          <div
            class="makeStyles-productTile-50"
          >
            <div
              class="makeStyles-productImage-49"
            />
            <div>
              <h6
                class="MuiTypography-root makeStyles-productName-47 MuiTypography-h6"
              >
                Product 8
              </h6>
              <div
                class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall"
                >
                  £8.00
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root makeStyles-productDescription-48 MuiTypography-body1"
            />
            &lt;Link /&gt;
          </div>
          <div
            class="makeStyles-productTile-55"
          >
            <div
              class="makeStyles-productImage-54"
            />
            <div>
              <h6
                class="MuiTypography-root makeStyles-productName-52 MuiTypography-h6"
              >
                Product 9
              </h6>
              <div
                class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall"
                >
                  £9.00
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root makeStyles-productDescription-53 MuiTypography-body1"
            />
            &lt;Link /&gt;
          </div>
          <div
            class="makeStyles-productTile-60"
          >
            <div
              class="makeStyles-productImage-59"
            />
            <div>
              <h6
                class="MuiTypography-root makeStyles-productName-57 MuiTypography-h6"
              >
                Product 10
              </h6>
              <div
                class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall"
                >
                  £10.00
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root makeStyles-productDescription-58 MuiTypography-body1"
            />
            &lt;Link /&gt;
          </div>
          <div
            class="makeStyles-productTile-65"
          >
            <div
              class="makeStyles-productImage-64"
            />
            <div>
              <h6
                class="MuiTypography-root makeStyles-productName-62 MuiTypography-h6"
              >
                Product 11
              </h6>
              <div
                class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall"
                >
                  £11.00
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root makeStyles-productDescription-63 MuiTypography-body1"
            />
            &lt;Link /&gt;
          </div>
          <div
            class="makeStyles-productTile-70"
          >
            <div
              class="makeStyles-productImage-69"
            />
            <div>
              <h6
                class="MuiTypography-root makeStyles-productName-67 MuiTypography-h6"
              >
                Product 12
              </h6>
              <div
                class="MuiChip-root makeStyles-chip-14 MuiChip-outlined MuiChip-sizeSmall"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall"
                >
                  £12.00
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root makeStyles-productDescription-68 MuiTypography-body1"
            />
            &lt;Link /&gt;
          </div>
        </div>
        <div
          class="makeStyles-leftArrow-7 makeStyles-arrow-6 makeStyles-disabled-5"
        />
        <div
          class="makeStyles-rightArrow-8 makeStyles-arrow-6"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
